src/engine-actions/prepare-swap.js

```diff
@@ -17,29 +8,53 @@ const SWAP_EXPIRY = '3600'
  * @type {String}
  * @default
  */
-const MEMO_PREFIX = 'sparkswap-swap-pivot:'
+const MEMO_PREFIX = 'sparkswap-swap-pivot'
 
 /**
  * Prepares for a swap in which this node is the counterparty to the intiating node
  *
- * @param {string} orderId - Order for which the swap is being executed
- * @param {string} swapHash        - swap hash that will be associated with the swap
- * @param {string} value   - Int64 string of the value of inbound currency
- * @returns {string} swapHash      Hash that will be translated
+ * @param {string} swapHash - swap hash that will be associated with the swap
+ * @param {string} value - Int64 string of the value of inbound currency
+ * @param {Date} expiryTime - absolute payment request expiry time
+ * @param {number} cltvExpiry - delta to use for the time-lock of the CLTV
+ *                              extended to the final hop (in seconds)
+ * @returns {string} paymentRequest - lightning invoice for a payment
  */
-async function prepareSwap (orderId, swapHash, value) {
+async function prepareSwap (swapHash, value, expiryTime, cltvExpiry) {
   this.logger.info(`Preparing swap for ${swapHash}`, { value })
+  // round up on number of blocks since a transaction that occurs right before
+  // the expiration time will get processed in the following block
+  const cltvExpiryBlocks = Math.ceil(cltvExpiry / this.secondsPerBlock)
+  // make prepareSwap idempotent by returning the existing invoice if one exists
+  try {
+    const {
+      paymentRequest,
+      value: invoiceValue,
+      cltvExpiry: invoiceCLTVExpiry
+    } = await lookupInvoice(swapHash, { client: this.client })
 
-  const params = {
-    memo: `${MEMO_PREFIX}${orderId}`,
-    value,
-    expiry: SWAP_EXPIRY,
-    hash: swapHash
-  }
-
-  const { paymentRequest } = await addHoldInvoice(params, { client: this.client })
+    if (invoiceValue !== value || invoiceCLTVExpiry !== cltvExpiryBlocks) {
+      // if the invoice doesn't match, we log an error and  prepare a new invoice
+      const hash = swapHash.toHex()
+      const message = `Invoice found for hash ${hash} but parameters are wrong`
+      this.logger.error(message)
+      throw new Error(message)
+    }
+    return paymentRequest
+  } catch (e) {
```

Comment:
Why do this in every error case? Aren't there some errors where we would *not* want to create a new invoice?

Reply:
Discussed in person: since there's no documentation for error types, we can't safely distinguish error types, and it isn't likely to make much difference anyway.